Add some more regression tests

JohnTitor opened the pull request:
This is another round of #77741. Tested with `debug-assertions=true` and it passed on my local.
Closes #70877
Closes #70944
Closes #71659
Closes #74816
Closes #75707
Closes #75983
(Skipped #63355 because I'm not sure about the error.)

nagisa commented on src/test/ui/unsized/issue-71659.rs:
```diff
@@ -0,0 +1,94 @@
+use std::rc::Rc;
+use std::sync::Arc;
+
+pub trait CastTo<T: ?Sized>: Unsize<T> {
```
Any reason this is not using the minimal version that's in the issue description (as opposed to what's in the playground?)

JohnTitor replied:
Ah, good point! I assumed that both examples were the same.

nagisa commented:
Overall LGTM, r=me but not instructing bors just in case you want to address the comment above, either way choice's yours.

---

As a side comment, these would have been easier to review if they were split into PRs for each of the `E-needs-test` issues.

JohnTitor commented:
Addressed the review via d80f93d507406375e302d9583e02bf19d1565cfe and rebased to make sure the diagnostics aren't changed.
@bors r=nagisa

> As a side comment, these would have been easier to review if they were split into PRs for each of the `E-needs-test` issues.

Fair enough, but in this case, #73905 fixed 10+ ICEs so it would have needed some more hustle :)

bors commented:
:pushpin: Commit d80f93d507406375e302d9583e02bf19d1565cfe has been approved by `nagisa`

<!-- @bors r=nagisa d80f93d507406375e302d9583e02bf19d1565cfe -->
<!-- homu: {"type":"Approved","sha":"d80f93d507406375e302d9583e02bf19d1565cfe","approver":"nagisa"} -->

bors commented:
:hourglass: Testing commit d80f93d507406375e302d9583e02bf19d1565cfe with merge cbc42a0f3eae5103888dbc10f3a7cf213977468c...
<!-- homu: {"type":"BuildStarted","head_sha":"d80f93d507406375e302d9583e02bf19d1565cfe","merge_sha":"cbc42a0f3eae5103888dbc10f3a7cf213977468c"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1270037922), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=37288)
Approved by: nagisa
Pushing cbc42a0f3eae5103888dbc10f3a7cf213977468c to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nagisa","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=37288","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1270037922"},"merge_sha":"cbc42a0f3eae5103888dbc10f3a7cf213977468c"} -->
